src/monodroid/jni/embedded-assemblies-zip.cc

```diff
@@ -0,0 +1,443 @@
+
+	uint16_t extra_field_length;
+	index = LH_EXTRA_LENGTH_OFFSET;
+	if (!zip_read_field (local_header, ZIP_LOCAL_LEN, index, extra_field_length)) {
+		log_fatal (LOG_DEFAULT, "Failed to read Local Header 'extra field length' field at offset %u", (local_header_offset + index));
+		return false;
+	}
+
+	data_start_offset = static_cast<uint32_t>(local_header_offset) + file_name_length + extra_field_length + ZIP_LOCAL_LEN;
+
+	return true;
+}
+
+bool
+EmbeddedAssemblies::zip_extract_cd_info (uint8_t* buf, size_t buf_len, uint32_t& cd_offset, uint32_t& cd_size, uint16_t& cd_entries)
+{
+	static constexpr size_t EOCD_TOTAL_ENTRIES_OFFSET = 10;
+	static constexpr size_t EOCD_CD_SIZE_OFFSET       = 12;
+	static constexpr size_t EOCD_CD_START_OFFSET      = 16;
+
+	if (buf_len < ZIP_EOCD_LEN) {
+		log_fatal (LOG_DEFAULT, "Buffer to short for EOCD");
+		return false;
+	}
+
+	if (!zip_read_field (buf, buf_len, EOCD_TOTAL_ENTRIES_OFFSET, cd_entries)) {
+		log_fatal (LOG_DEFAULT, "Failed to read EOCD 'total number of entries' field");
+		return false;
+	}
+
+	if (!zip_read_field (buf, buf_len, EOCD_CD_START_OFFSET, cd_offset)) {
+		log_fatal (LOG_DEFAULT, "Failed to read EOCD 'central directory size' field");
+		return false;
+	}
+
+	if (!zip_read_field (buf, buf_len, EOCD_CD_SIZE_OFFSET, cd_size)) {
+		log_fatal (LOG_DEFAULT, "Failed to read EOCD 'central directory offset' field");
+		return false;
+	}
+
+	return true;
+}
+
+bool
+EmbeddedAssemblies::zip_ensure_valid_params (uint8_t* buf, size_t buf_len, size_t index, size_t to_read)
+{
+	assert (buf != nullptr);
+	if (index + to_read > buf_len) {
+		log_fatal (LOG_DEFAULT, "Buffer too short to read %u bytes of data", to_read);
+		return false;
+	}
+
+	return true;
+}
+
+bool
+EmbeddedAssemblies::zip_read_field (uint8_t* buf, size_t buf_len, size_t index, uint16_t& u)
+{
+	if (!zip_ensure_valid_params (buf, buf_len, index, sizeof (u)))
+		return false;
+
+	u = static_cast<uint16_t> (buf [index + 1] << 8) |
+		static_cast<uint16_t> (buf [index]);
+
+	return true;
+}
+
+bool
+EmbeddedAssemblies::zip_read_field (uint8_t* buf, size_t buf_len, size_t index, uint32_t& u)
+{
+	if (!zip_ensure_valid_params (buf, buf_len, index, sizeof (u)))
+		return false;
+
+	u = (static_cast<uint32_t> (buf [index + 3]) << 24) |
+		(static_cast<uint32_t> (buf [index + 2]) << 16) |
+		(static_cast<uint32_t> (buf [index + 1]) << 8)  |
+		(static_cast<uint32_t> (buf [index + 0]));
+
+	return true;
+}
+
+bool
+EmbeddedAssemblies::zip_read_field (uint8_t* buf, size_t buf_len, size_t index, uint8_t (&sig)[4])
+{
+	static constexpr size_t sig_size = sizeof(sig);
+
+	if (!zip_ensure_valid_params (buf, buf_len, index, sig_size))
+		return false;
+
+	memcpy (sig, buf + index, sig_size);
+	return true;
+}
+
+bool
+EmbeddedAssemblies::zip_read_field (uint8_t* buf, size_t buf_len, size_t index, size_t count, char*& characters)
+{
+	if (!zip_ensure_valid_params (buf, buf_len, index, count))
+		return false;
+
+	characters = new char[count + 1];
```

Comment:
Should this use `ADD_WITH_OVERFLOW_CHECK()` in case `index` is some "unreasonable" value so that we don't overflow?